### PR TITLE
Enable the easy download of the deployment.tar.gz

### PR DESCRIPTION
**Pull request by dbogunowicz**
Reverting to the state of #375 in order to enable, in the simplest way possible, the fetching of only `deployment.tar.gz`.
The crux of this PR is adding `self.deployment_tar` attribute to the `Model` and allowing it to be correctly initialized not only from the stub but also from the local directory.
Added unit tests that test this feature.

Manual tests:

```python
path_where_models_are = "/home/ubuntu/.cache/sparsezoo/neuralmagic/codegen_mono-350m-bigpython_bigquery_thepile-pruned50_quantized"
if os.path.exists(path_where_models_are):
    shutil.rmtree(path_where_models_are)
    print(f"The path {path_where_models_are} has been cleaned up")
else:
    print(f"The path {path_where_models_are} does not exist")


print("1: Creating the deployment from stub")
model = Model("zoo:codegen_mono-350m-bigpython_bigquery_thepile-pruned50_quantized")
# local path to the `deployment.tar.gz` should be None (we have not downloaded anything yet)
print("Deployment directory tar path (before download): ", model.deployment_tar._path)
# grab the deployment_dir_path (this downloads the deployment.tar.gz and unpacks it)
deployment_dir_path = model.deployment_directory_path

# all should be pointing to `directory`
assert deployment_dir_path == model.deployment_tar.path == model.deployment.path
print("Deployment directory tar path: ", model.deployment_tar._path)
print("Deployment directory path: ", model.deployment.path)
print(f"deployment_dir_path: {deployment_dir_path}")

# all should display `directory` files
print("Contents of the deployment tar directory: ", os.listdir(model.deployment_tar.path))
print("Contents of the deployment directory: ", os.listdir(model.deployment.path))

# should only contain `deployment` and `deployment.tar.gz`
print("Contents of the model directory: ", os.listdir(model._path))

print("2: Creating the deployment from local path")
model = Model("zoo:codegen_mono-350m-bigpython_bigquery_thepile-pruned50_quantized")
print("Deployment directory tar path: ", model.deployment_tar.path)
deployment_dir_path = model.deployment_directory_path

assert deployment_dir_path == model.deployment_tar.path
print("Deployment directory tar path (after unzipping): ", model.deployment_tar._path)
print("Deployment directory path: ", model.deployment.path)
print(f"deployment_dir_path: {deployment_dir_path}")
print("Contents of the deployment tar directory: ", os.listdir(model.deployment_tar.path))
print("Contents of the deployment directory: ", os.listdir(model.deployment.path))
print("Contents of the model directory: ", os.listdir(model._path))
```

out:

```bash
The path /home/ubuntu/.cache/sparsezoo/neuralmagic/codegen_mono-350m-bigpython_bigquery_thepile-pruned50_quantized has been cleaned up
1: Creating the deployment from stub
Deployment directory tar path (before download):  None
Downloading (…)ed/deployment.tar.gz: 100%|██████████| 265M/265M [00:23<00:00, 12.1MB/s]
Deployment directory tar path:  /home/ubuntu/.cache/sparsezoo/neuralmagic/codegen_mono-350m-bigpython_bigquery_thepile-pruned50_quantized/deployment
Deployment directory path:  /home/ubuntu/.cache/sparsezoo/neuralmagic/codegen_mono-350m-bigpython_bigquery_thepile-pruned50_quantized/deployment
deployment_dir_path: /home/ubuntu/.cache/sparsezoo/neuralmagic/codegen_mono-350m-bigpython_bigquery_thepile-pruned50_quantized/deployment
Contents of the deployment tar directory:  ['vocab.json', 'model.onnx', 'config.json', 'merges.txt', 'tokenizer_config.json', 'special_tokens_map.json', 'tokenizer.json']
Contents of the deployment directory:  ['vocab.json', 'model.onnx', 'config.json', 'merges.txt', 'tokenizer_config.json', 'special_tokens_map.json', 'tokenizer.json']
Contents of the model directory:  ['deployment', 'deployment.tar.gz']
2: Creating the deployment from local path
Deployment directory tar path:  /home/ubuntu/.cache/sparsezoo/neuralmagic/codegen_mono-350m-bigpython_bigquery_thepile-pruned50_quantized/deployment.tar.gz
Deployment directory tar path (after unzipping):  /home/ubuntu/.cache/sparsezoo/neuralmagic/codegen_mono-350m-bigpython_bigquery_thepile-pruned50_quantized/deployment
Deployment directory path:  /home/ubuntu/.cache/sparsezoo/neuralmagic/codegen_mono-350m-bigpython_bigquery_thepile-pruned50_quantized/deployment
deployment_dir_path: /home/ubuntu/.cache/sparsezoo/neuralmagic/codegen_mono-350m-bigpython_bigquery_thepile-pruned50_quantized/deployment
Contents of the deployment tar directory:  ['vocab.json', 'model.onnx', 'config.json', 'merges.txt', 'tokenizer_config.json', 'special_tokens_map.json', 'tokenizer.json']
Contents of the deployment directory:  ['vocab.json', 'model.onnx', 'config.json', 'merges.txt', 'tokenizer_config.json', 'special_tokens_map.json', 'tokenizer.json']
Contents of the model directory:  ['deployment', 'deployment.tar.gz']
```

### CODE REVIEW

File: src/sparsezoo/model/model.py [hunk 1]
```diff
@@ -259,6 +257,8 @@ def path(self) -> str:
 
         # extract onnx_model if it is a tar.gz
         self.onnx_model.path
+        # extract deployment directory if it is a tar.gz
+        self.deployment_directory_path
```
Comment (dbogunowicz) [inline src/sparsezoo/model/model.py:261]: We need this line of code if we wish to untar `deployment.tar.gz` after we execute `Model(stub).download()` (see the example in the PR description)

File: tests/sparsezoo/model/test_model.py [hunk 2]
```diff
@@ -222,7 +223,7 @@ def _add_mock_files(directory_path: str, clone_sample_outputs: bool):
         os.makedirs(onnx_folder_dir)
         for opset in range(1, 3):
             shutil.copyfile(
-                os.path.join(directory_path, "deployment", "model.onnx"),
```
[the remark this answers is not in the document]
Comment (dbogunowicz) [reply tests/sparsezoo/model/test_model.py:225]: model.path (download())
model.deployment.path -> point to tar.gz

File: src/sparsezoo/model/model.py [hunk 3]
```diff
@@ -739,11 +738,35 @@ def _sample_outputs_list_to_dict(
     def _get_directory_from_loose_api_files(
         files, directory_class, display_name, parent_directory, **kwargs
     ):
-        # fetch all the loose files that belong to the directory (use `file_type` key
+        # fetch all the loose files that belong to the directory (use `display_name` key
         # from the `request_json` for proper mapping)
-        files = [
-            file_dict for file_dict in files if file_dict["file_type"] == display_name
+        files_by_display_name = [
```
Comment (dbogunowicz) [inline src/sparsezoo/model/model.py:743]: call model.deployment_directory_path:
 -> 1. No local deployment or deployment.tar.gz --> download only deployment.tar.gz and untar 
 -> 2. There is local deployment extracted > simply create model.directory: SelectDirectory(files=...)
 -> 3. There is a local deployment archived-> untar and create model.directory: SelectedDirectory(files=...)

File: src/sparsezoo/model/model.py [hunk 4]
```diff
@@ -233,11 +238,11 @@ def deployment_directory_path(self) -> str:
...
         # trigger initial download if not downloaded
-        self.deployment.path
-        if self.deployment.is_archive:
-            self.deployment.unzip()
+        self.deployment_tar.path
+        if self.deployment_tar.is_archive:
+            self.deployment_tar.unzip()
 
-        return self.deployment.path
+        return self.deployment_tar.path
```
Comment (bfineran) [inline src/sparsezoo/model/model.py:245]: shouldn't we return `deployment.path` now?

File: tests/sparsezoo/model/test_model.py [hunk 5]
```diff
@@ -329,6 +331,55 @@ def test_model_gz_extraction_from_local_files(stub: str):
...
+    expected_deployment_files = ["model.onnx"]
+
+    model = Model(stub, tmpdir)
+    assert model.deployment_tar.is_archive
+    # download and extract deployment tar
+    deployment_dir_path = model.deployment_directory_path
+
+    # deployment and deployment_tar should be point to the same files
+    assert deployment_dir_path == model.deployment_tar.path == model.deployment.path
+    # make sure that the model contains expectedd files
```
Comment (dbogunowicz) [inline tests/sparsezoo/model/test_model.py:351]:
```suggestion
    # make sure that the model contains expected files
```

**Review by bfineran — APPROVED** [review]

**Review by rahul-tuli — APPROVED** [review]

File: src/sparsezoo/model/model.py [hunk 6]
```diff
@@ -138,12 +137,26 @@ def __init__(self, source: str, download_path: Optional[str] = None):
...
+            allow_multiple_outputs=True,
+        )
+
+        if isinstance(self.deployment, list):
+            # if there are multiple deployment directories
+            # (this may happen due to the presence of both
+            # - deployment directory
+            # - deployment.tar.gz file
+            # we need to choose one (they are identical)
+            self.deployment = self.deployment[0]
```
[the remark this answers is not in the document]
Comment (rahul-tuli) [reply src/sparsezoo/model/model.py:154]: Saw the other PR kindly ignore this comment

**Review by bfineran — APPROVED** [review]

**Pull request merged**